File: ai/tetris/block.py
import numpy as np
from tetris.shape import Shape


class Block:
    """
    A class to represent the blocks which fall in the game. Each has a shape and
    starts at a rotation of 0.
    """
    def __init__(self, shape: Shape, rot: int = 0, x: int = 0, y: int = 0) -> None:
        self.shape = shape
        self.rot = rot
        self.x = x
        self.y = y
        self.width = shape.width
        self.height = shape.height

    def matrix(self) -> np.ndarray:
        return self.shape.rots[self.rot]


def rot_cw(block: Block) -> Block:
    """
    Returns a new block, rotated 90 degrees clockwise.
    The original block is unchanged.
    """
    return Block(
        shape=block.shape,
        rot=(block.rot + 1) % 4,
        x=block.x,
        y=block.y,
    )


def rot_ccw(block: Block) -> Block:
    """
    Returns a new block, rotated 90 degrees counter-clockwise.
    The original block is unchanged.
    """
    return Block(
        shape=block.shape,
        rot=(block.rot - 1) % 4,
        x=block.x,
        y=block.y,
    )


def down(block: Block) -> Block:
    """
    Returns a new block, shifted 1 y down.
    The original block is unchanged.
    """
    return Block(
        shape=block.shape,
        rot=block.rot,
        x=block.x,
        y=block.y + 1,
    )


def left(block: Block) -> Block:
    """
    Returns a new block, shifted 1 x down (left).
    The original block is unchanged.
    """
    return Block(
        shape=block.shape,
        rot=block.rot,
        x=block.x - 1,
        y=block.y,
    )


def right(block: Block) -> Block:
    """
    Returns a new block, shifted 1 x up (right).
    The original block is unchanged.
    """
    return Block(
        shape=block.shape,
        rot=block.rot,
        x=block.x + 1,
        y=block.y,
    )

File: ui/src/block.js
require("utils");
require("shape");
require("colors");

class Block {
	constructor(shape, color, rot = 0, x = 0, y = 0) {
		this.shape = shape;
		this.color = color;
		this.rot = rot;
		this.x = x;
		this.y = y;
		this.width = shape.width;
		this.height = shape.height;
	}

	matrix() {
		return this.shape.rots[this.rot];
	}

	map_rot(mapper) {
		return new Block(
			this.shape,
			this.color,
			mapper(this.rot),
			this.x,
			this.y
		)
	}

	map_loc(mapper) {
		xy = mapper(this.x, this.y);
		return new Block(
			this.shape,
			this.color,
			this.rot,
			xy.first,
			xy.second
		)
	}

	rot_cw() {
		return this.map_rot(rot => (rot + 1) % 4);
	}

	rot_ccw() {
		return this.map_rot(rot => (rot - 1) % 4);
	}

	down() {
		this.map_loc((x, y) => new Pair(x, y + 1));
	}

	left() {
		this.map_loc((x, y) => new Pair(x - 1, y));
	}

	right() {
		this.map_loc((x, y) => new Pair(x + 1, y));
	}
}

file_loaded("block");
